lightningd/peer_control.c

```diff
@@ -1339,6 +1339,102 @@ static const struct json_command getinfo_command = {
 };
 AUTODATA(json_command, &getinfo_command);
 
+
+static struct command_result *json_setchannelfee(struct command *cmd,
+					 const char *buffer,
+					 const jsmntok_t *obj UNNEEDED,
+					 const jsmntok_t *params)
+{
+	struct json_stream *response;
+	const jsmntok_t *idtok;
+	struct peer *peer;
+	struct channel *channel;
+	struct channel_id cid;
+	struct command_result *res;
+	struct amount_msat *base, base_default;
+	u32 *ppm, ppm_default;
+
+	/* Use daemon config values base/ppm as default */
+	amount_msat_from_u32(&base_default, cmd->ld->config.fee_base);
+	ppm_default = cmd->ld->config.fee_per_satoshi;
+
+	/* Parse the JSON command */
+	if (!param(cmd, buffer, params,
+			p_req("id", param_tok, &idtok),
+			p_opt_def("base", param_msat, &base, base_default),
+			p_opt_def("ppm", param_number, &ppm, ppm_default),
+			NULL))
+		return command_param_failed();
+
+	/* Check base for u32 overflow */
+	if (amount_msat_overflow(*base))
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				"Value 'base' exceeds u32 max");
+
+	/* Find the channel */
+	peer = peer_from_json(cmd->ld, buffer, idtok);
+	if (peer) {
+		channel = peer_active_channel(peer);
+		if (!channel)
+			return command_fail(cmd, LIGHTNINGD,
+					"Could not find active channel of peer with that id");
+	} else {
+		res = command_find_channel(cmd, buffer, idtok, &channel);
+		if (res)
+			return res;
+		/* check channel is found and in valid state */
+		if (!channel)
+			return command_fail(cmd, LIGHTNINGD,
+					"Could not find channel with that id");
+	}
+
+	if (channel->state != CHANNELD_NORMAL)
+		return command_fail(cmd, LIGHTNINGD,
+				"Channel is not in state CHANNELD_NORMAL");
+
+	/* set, notify and save new values */
+	channel->feerate_base = amount_msat_to_u32(*base);
+	channel->feerate_ppm = *ppm;
+
+	/* tell channeld to make a send_channel_update */
+	if (channel->owner && streq(channel->owner->name, "lightning_channeld"))
+		subd_send_msg(channel->owner,
+				take(towire_channel_specific_feerates(channel, amount_msat_to_u32(*base), *ppm)));
+
+	/* save values to database */
+	wallet_channel_save(cmd->ld->wallet, channel);
+
+	/* get channel id */
+	derive_channel_id(&cid, &channel->funding_txid, channel->funding_outnum);
+
+	/* write response */
+	response = json_stream_success(cmd);
+	json_object_start(response, NULL);
+	json_add_num(response, "base", amount_msat_to_u32(*base));
+	json_add_num(response, "ppm", *ppm);
+	json_add_string(response, "channel_id",
+			type_to_string(tmpctx, struct channel_id, &cid));
+	if (peer)
```

Comment:
Can we actually ever get here without having a `peer`?

Reply:
addressed